Traceback (most recent call last):
  File "/Users/fbi005/miniconda3/envs/sta2003_nb/lib/python3.10/site-packages/jupyter_cache/executors/utils.py", line 58, in single_nb_execution
    executenb(
  File "/Users/fbi005/miniconda3/envs/sta2003_nb/lib/python3.10/site-packages/nbclient/client.py", line 1319, in execute
    return NotebookClient(nb=nb, resources=resources, km=km, **kwargs).execute()
  File "/Users/fbi005/miniconda3/envs/sta2003_nb/lib/python3.10/site-packages/jupyter_core/utils/__init__.py", line 165, in wrapped
    return loop.run_until_complete(inner)
  File "/Users/fbi005/miniconda3/envs/sta2003_nb/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/Users/fbi005/miniconda3/envs/sta2003_nb/lib/python3.10/site-packages/nbclient/client.py", line 709, in async_execute
    await self.async_execute_cell(
  File "/Users/fbi005/miniconda3/envs/sta2003_nb/lib/python3.10/site-packages/nbclient/client.py", line 1062, in async_execute_cell
    await self._check_raise_for_error(cell, cell_index, exec_reply)
  File "/Users/fbi005/miniconda3/envs/sta2003_nb/lib/python3.10/site-packages/nbclient/client.py", line 918, in _check_raise_for_error
    raise CellExecutionError.from_cell_and_msg(cell, exec_reply_content)
nbclient.exceptions.CellExecutionError: An error occurred while executing the following cell:
------------------
co2 = sm.datasets.get_rdataset("co2", "datasets").data
print(co2.head())

# Convert decimal year to pandas datetime
def convert_decimal_year_to_datetime(decimal_years):
    dates = [(pd.to_datetime(f'{int(year)}-01-01') + pd.to_timedelta((year - int(year)) * 365.25, unit='D')).date() 
             for year in decimal_years]
    return dates
co2['time'] = convert_decimal_year_to_datetime(co2['time'])

# Convert the column ds to datetime
co2['time'] = pd.to_datetime(co2['time'])
print("\nConverted:\n------------------\n", co2.head())

# Resample to monthly frequency based on the ds column
co2 = co2.resample('MS', on='time').mean().reset_index()

# Replace NaN with the mean of the previous and next value
co2['value'] = co2['value'].interpolate()
print("\nResampled:\n------------------\n", co2.head())
------------------


[0;31m---------------------------------------------------------------------------[0m
[0;31mgaierror[0m                                  Traceback (most recent call last)
File [0;32m~/miniconda3/envs/sta2003_nb/lib/python3.10/urllib/request.py:1348[0m, in [0;36mAbstractHTTPHandler.do_open[0;34m(self, http_class, req, **http_conn_args)[0m
[1;32m   1347[0m [38;5;28;01mtry[39;00m:
[0;32m-> 1348[0m     [43mh[49m[38;5;241;43m.[39;49m[43mrequest[49m[43m([49m[43mreq[49m[38;5;241;43m.[39;49m[43mget_method[49m[43m([49m[43m)[49m[43m,[49m[43m [49m[43mreq[49m[38;5;241;43m.[39;49m[43mselector[49m[43m,[49m[43m [49m[43mreq[49m[38;5;241;43m.[39;49m[43mdata[49m[43m,[49m[43m [49m[43mheaders[49m[43m,[49m
[1;32m   1349[0m [43m              [49m[43mencode_chunked[49m[38;5;241;43m=[39;49m[43mreq[49m[38;5;241;43m.[39;49m[43mhas_header[49m[43m([49m[38;5;124;43m'[39;49m[38;5;124;43mTransfer-encoding[39;49m[38;5;124;43m'[39;49m[43m)[49m[43m)[49m
[1;32m   1350[0m [38;5;28;01mexcept[39;00m [38;5;167;01mOSError[39;00m [38;5;28;01mas[39;00m err: [38;5;66;03m# timeout error[39;00m

File [0;32m~/miniconda3/envs/sta2003_nb/lib/python3.10/http/client.py:1283[0m, in [0;36mHTTPConnection.request[0;34m(self, method, url, body, headers, encode_chunked)[0m
[1;32m   1282[0m [38;5;250m[39m[38;5;124;03m"""Send a complete request to the server."""[39;00m
[0;32m-> 1283[0m [38;5;28;43mself[39;49m[38;5;241;43m.[39;49m[43m_send_request[49m[43m([49m[43mmethod[49m[43m,[49m[43m [49m[43murl[49m[43m,[49m[43m [49m[43mbody[49m[43m,[49m[43m [49m[43mheaders[49m[43m,[49m[43m [49m[43mencode_chunked[49m[43m)[49m

File [0;32m~/miniconda3/envs/sta2003_nb/lib/python3.10/http/client.py:1329[0m, in [0;36mHTTPConnection._send_request[0;34m(self, method, url, body, headers, encode_chunked)[0m
[1;32m   1328[0m     body [38;5;241m=[39m _encode(body, [38;5;124m'[39m[38;5;124mbody[39m[38;5;124m'[39m)
[0;32m-> 1329[0m [38;5;28;43mself[39;49m[38;5;241;43m.[39;49m[43mendheaders[49m[43m([49m[43mbody[49m[43m,[49m[43m [49m[43mencode_chunked[49m[38;5;241;43m=[39;49m[43mencode_chunked[49m[43m)[49m

File [0;32m~/miniconda3/envs/sta2003_nb/lib/python3.10/http/client.py:1278[0m, in [0;36mHTTPConnection.endheaders[0;34m(self, message_body, encode_chunked)[0m
[1;32m   1277[0m     [38;5;28;01mraise[39;00m CannotSendHeader()
[0;32m-> 1278[0m [38;5;28;43mself[39;49m[38;5;241;43m.[39;49m[43m_send_output[49m[43m([49m[43mmessage_body[49m[43m,[49m[43m [49m[43mencode_chunked[49m[38;5;241;43m=[39;49m[43mencode_chunked[49m[43m)[49m

File [0;32m~/miniconda3/envs/sta2003_nb/lib/python3.10/http/client.py:1038[0m, in [0;36mHTTPConnection._send_output[0;34m(self, message_body, encode_chunked)[0m
[1;32m   1037[0m [38;5;28;01mdel[39;00m [38;5;28mself[39m[38;5;241m.[39m_buffer[:]
[0;32m-> 1038[0m [38;5;28;43mself[39;49m[38;5;241;43m.[39;49m[43msend[49m[43m([49m[43mmsg[49m[43m)[49m
[1;32m   1040[0m [38;5;28;01mif[39;00m message_body [38;5;129;01mis[39;00m [38;5;129;01mnot[39;00m [38;5;28;01mNone[39;00m:
[1;32m   1041[0m 
[1;32m   1042[0m     [38;5;66;03m# create a consistent interface to message_body[39;00m

File [0;32m~/miniconda3/envs/sta2003_nb/lib/python3.10/http/client.py:976[0m, in [0;36mHTTPConnection.send[0;34m(self, data)[0m
[1;32m    975[0m [38;5;28;01mif[39;00m [38;5;28mself[39m[38;5;241m.[39mauto_open:
[0;32m--> 976[0m     [38;5;28;43mself[39;49m[38;5;241;43m.[39;49m[43mconnect[49m[43m([49m[43m)[49m
[1;32m    977[0m [38;5;28;01melse[39;00m:

File [0;32m~/miniconda3/envs/sta2003_nb/lib/python3.10/http/client.py:1448[0m, in [0;36mHTTPSConnection.connect[0;34m(self)[0m
[1;32m   1446[0m [38;5;124m"[39m[38;5;124mConnect to a host on a given (SSL) port.[39m[38;5;124m"[39m
[0;32m-> 1448[0m [38;5;28;43msuper[39;49m[43m([49m[43m)[49m[38;5;241;43m.[39;49m[43mconnect[49m[43m([49m[43m)[49m
[1;32m   1450[0m [38;5;28;01mif[39;00m [38;5;28mself[39m[38;5;241m.[39m_tunnel_host:

File [0;32m~/miniconda3/envs/sta2003_nb/lib/python3.10/http/client.py:942[0m, in [0;36mHTTPConnection.connect[0;34m(self)[0m
[1;32m    941[0m sys[38;5;241m.[39maudit([38;5;124m"[39m[38;5;124mhttp.client.connect[39m[38;5;124m"[39m, [38;5;28mself[39m, [38;5;28mself[39m[38;5;241m.[39mhost, [38;5;28mself[39m[38;5;241m.[39mport)
[0;32m--> 942[0m [38;5;28mself[39m[38;5;241m.[39msock [38;5;241m=[39m [38;5;28;43mself[39;49m[38;5;241;43m.[39;49m[43m_create_connection[49m[43m([49m
[1;32m    943[0m [43m    [49m[43m([49m[38;5;28;43mself[39;49m[38;5;241;43m.[39;49m[43mhost[49m[43m,[49m[38;5;28;43mself[39;49m[38;5;241;43m.[39;49m[43mport[49m[43m)[49m[43m,[49m[43m [49m[38;5;28;43mself[39;49m[38;5;241;43m.[39;49m[43mtimeout[49m[43m,[49m[43m [49m[38;5;28;43mself[39;49m[38;5;241;43m.[39;49m[43msource_address[49m[43m)[49m
[1;32m    944[0m [38;5;66;03m# Might fail in OSs that don't implement TCP_NODELAY[39;00m

File [0;32m~/miniconda3/envs/sta2003_nb/lib/python3.10/socket.py:836[0m, in [0;36mcreate_connection[0;34m(address, timeout, source_address)[0m
[1;32m    835[0m err [38;5;241m=[39m [38;5;28;01mNone[39;00m
[0;32m--> 836[0m [38;5;28;01mfor[39;00m res [38;5;129;01min[39;00m [43mgetaddrinfo[49m[43m([49m[43mhost[49m[43m,[49m[43m [49m[43mport[49m[43m,[49m[43m [49m[38;5;241;43m0[39;49m[43m,[49m[43m [49m[43mSOCK_STREAM[49m[43m)[49m:
[1;32m    837[0m     af, socktype, proto, canonname, sa [38;5;241m=[39m res

File [0;32m~/miniconda3/envs/sta2003_nb/lib/python3.10/socket.py:967[0m, in [0;36mgetaddrinfo[0;34m(host, port, family, type, proto, flags)[0m
[1;32m    966[0m addrlist [38;5;241m=[39m []
[0;32m--> 967[0m [38;5;28;01mfor[39;00m res [38;5;129;01min[39;00m [43m_socket[49m[38;5;241;43m.[39;49m[43mgetaddrinfo[49m[43m([49m[43mhost[49m[43m,[49m[43m [49m[43mport[49m[43m,[49m[43m [49m[43mfamily[49m[43m,[49m[43m [49m[38;5;28;43mtype[39;49m[43m,[49m[43m [49m[43mproto[49m[43m,[49m[43m [49m[43mflags[49m[43m)[49m:
[1;32m    968[0m     af, socktype, proto, canonname, sa [38;5;241m=[39m res

[0;31mgaierror[0m: [Errno 8] nodename nor servname provided, or not known

During handling of the above exception, another exception occurred:

[0;31mURLError[0m                                  Traceback (most recent call last)
Cell [0;32mIn[68], line 1[0m
[0;32m----> 1[0m co2 [38;5;241m=[39m [43msm[49m[38;5;241;43m.[39;49m[43mdatasets[49m[38;5;241;43m.[39;49m[43mget_rdataset[49m[43m([49m[38;5;124;43m"[39;49m[38;5;124;43mco2[39;49m[38;5;124;43m"[39;49m[43m,[49m[43m [49m[38;5;124;43m"[39;49m[38;5;124;43mdatasets[39;49m[38;5;124;43m"[39;49m[43m)[49m[38;5;241m.[39mdata
[1;32m      2[0m [38;5;28mprint[39m(co2[38;5;241m.[39mhead())
[1;32m      4[0m [38;5;66;03m# Convert decimal year to pandas datetime[39;00m

File [0;32m~/miniconda3/envs/sta2003_nb/lib/python3.10/site-packages/statsmodels/datasets/utils.py:237[0m, in [0;36mget_rdataset[0;34m(dataname, package, cache)[0m
[1;32m    234[0m docs_base_url [38;5;241m=[39m ([38;5;124m"[39m[38;5;124mhttps://raw.githubusercontent.com/vincentarelbundock/Rdatasets/[39m[38;5;124m"[39m
[1;32m    235[0m                  [38;5;124m"[39m[38;5;124mmaster/doc/[39m[38;5;124m"[39m[38;5;241m+[39mpackage[38;5;241m+[39m[38;5;124m"[39m[38;5;124m/rst/[39m[38;5;124m"[39m)
[1;32m    236[0m cache [38;5;241m=[39m _get_cache(cache)
[0;32m--> 237[0m data, from_cache [38;5;241m=[39m [43m_get_data[49m[43m([49m[43mdata_base_url[49m[43m,[49m[43m [49m[43mdataname[49m[43m,[49m[43m [49m[43mcache[49m[43m)[49m
[1;32m    238[0m data [38;5;241m=[39m read_csv(data, index_col[38;5;241m=[39m[38;5;241m0[39m)
[1;32m    239[0m data [38;5;241m=[39m _maybe_reset_index(data)

File [0;32m~/miniconda3/envs/sta2003_nb/lib/python3.10/site-packages/statsmodels/datasets/utils.py:166[0m, in [0;36m_get_data[0;34m(base_url, dataname, cache, extension)[0m
[1;32m    164[0m url [38;5;241m=[39m base_url [38;5;241m+[39m (dataname [38;5;241m+[39m [38;5;124m"[39m[38;5;124m.[39m[38;5;132;01m%s[39;00m[38;5;124m"[39m) [38;5;241m%[39m extension
[1;32m    165[0m [38;5;28;01mtry[39;00m:
[0;32m--> 166[0m     data, from_cache [38;5;241m=[39m [43m_urlopen_cached[49m[43m([49m[43murl[49m[43m,[49m[43m [49m[43mcache[49m[43m)[49m
[1;32m    167[0m [38;5;28;01mexcept[39;00m HTTPError [38;5;28;01mas[39;00m err:
[1;32m    168[0m     [38;5;28;01mif[39;00m [38;5;124m'[39m[38;5;124m404[39m[38;5;124m'[39m [38;5;129;01min[39;00m [38;5;28mstr[39m(err):

File [0;32m~/miniconda3/envs/sta2003_nb/lib/python3.10/site-packages/statsmodels/datasets/utils.py:157[0m, in [0;36m_urlopen_cached[0;34m(url, cache)[0m
[1;32m    155[0m [38;5;66;03m# not using the cache or did not find it in cache[39;00m
[1;32m    156[0m [38;5;28;01mif[39;00m [38;5;129;01mnot[39;00m from_cache:
[0;32m--> 157[0m     data [38;5;241m=[39m [43murlopen[49m[43m([49m[43murl[49m[43m,[49m[43m [49m[43mtimeout[49m[38;5;241;43m=[39;49m[38;5;241;43m3[39;49m[43m)[49m[38;5;241m.[39mread()
[1;32m    158[0m     [38;5;28;01mif[39;00m cache [38;5;129;01mis[39;00m [38;5;129;01mnot[39;00m [38;5;28;01mNone[39;00m:  [38;5;66;03m# then put it in the cache[39;00m
[1;32m    159[0m         _cache_it(data, cache_path)

File [0;32m~/miniconda3/envs/sta2003_nb/lib/python3.10/urllib/request.py:216[0m, in [0;36murlopen[0;34m(url, data, timeout, cafile, capath, cadefault, context)[0m
[1;32m    214[0m [38;5;28;01melse[39;00m:
[1;32m    215[0m     opener [38;5;241m=[39m _opener
[0;32m--> 216[0m [38;5;28;01mreturn[39;00m [43mopener[49m[38;5;241;43m.[39;49m[43mopen[49m[43m([49m[43murl[49m[43m,[49m[43m [49m[43mdata[49m[43m,[49m[43m [49m[43mtimeout[49m[43m)[49m

File [0;32m~/miniconda3/envs/sta2003_nb/lib/python3.10/urllib/request.py:519[0m, in [0;36mOpenerDirector.open[0;34m(self, fullurl, data, timeout)[0m
[1;32m    516[0m     req [38;5;241m=[39m meth(req)
[1;32m    518[0m sys[38;5;241m.[39maudit([38;5;124m'[39m[38;5;124murllib.Request[39m[38;5;124m'[39m, req[38;5;241m.[39mfull_url, req[38;5;241m.[39mdata, req[38;5;241m.[39mheaders, req[38;5;241m.[39mget_method())
[0;32m--> 519[0m response [38;5;241m=[39m [38;5;28;43mself[39;49m[38;5;241;43m.[39;49m[43m_open[49m[43m([49m[43mreq[49m[43m,[49m[43m [49m[43mdata[49m[43m)[49m
[1;32m    521[0m [38;5;66;03m# post-process response[39;00m
[1;32m    522[0m meth_name [38;5;241m=[39m protocol[38;5;241m+[39m[38;5;124m"[39m[38;5;124m_response[39m[38;5;124m"[39m

File [0;32m~/miniconda3/envs/sta2003_nb/lib/python3.10/urllib/request.py:536[0m, in [0;36mOpenerDirector._open[0;34m(self, req, data)[0m
[1;32m    533[0m     [38;5;28;01mreturn[39;00m result
[1;32m    535[0m protocol [38;5;241m=[39m req[38;5;241m.[39mtype
[0;32m--> 536[0m result [38;5;241m=[39m [38;5;28;43mself[39;49m[38;5;241;43m.[39;49m[43m_call_chain[49m[43m([49m[38;5;28;43mself[39;49m[38;5;241;43m.[39;49m[43mhandle_open[49m[43m,[49m[43m [49m[43mprotocol[49m[43m,[49m[43m [49m[43mprotocol[49m[43m [49m[38;5;241;43m+[39;49m
[1;32m    537[0m [43m                          [49m[38;5;124;43m'[39;49m[38;5;124;43m_open[39;49m[38;5;124;43m'[39;49m[43m,[49m[43m [49m[43mreq[49m[43m)[49m
[1;32m    538[0m [38;5;28;01mif[39;00m result:
[1;32m    539[0m     [38;5;28;01mreturn[39;00m result

File [0;32m~/miniconda3/envs/sta2003_nb/lib/python3.10/urllib/request.py:496[0m, in [0;36mOpenerDirector._call_chain[0;34m(self, chain, kind, meth_name, *args)[0m
[1;32m    494[0m [38;5;28;01mfor[39;00m handler [38;5;129;01min[39;00m handlers:
[1;32m    495[0m     func [38;5;241m=[39m [38;5;28mgetattr[39m(handler, meth_name)
[0;32m--> 496[0m     result [38;5;241m=[39m [43mfunc[49m[43m([49m[38;5;241;43m*[39;49m[43margs[49m[43m)[49m
[1;32m    497[0m     [38;5;28;01mif[39;00m result [38;5;129;01mis[39;00m [38;5;129;01mnot[39;00m [38;5;28;01mNone[39;00m:
[1;32m    498[0m         [38;5;28;01mreturn[39;00m result

File [0;32m~/miniconda3/envs/sta2003_nb/lib/python3.10/urllib/request.py:1391[0m, in [0;36mHTTPSHandler.https_open[0;34m(self, req)[0m
[1;32m   1390[0m [38;5;28;01mdef[39;00m[38;5;250m [39m[38;5;21mhttps_open[39m([38;5;28mself[39m, req):
[0;32m-> 1391[0m     [38;5;28;01mreturn[39;00m [38;5;28;43mself[39;49m[38;5;241;43m.[39;49m[43mdo_open[49m[43m([49m[43mhttp[49m[38;5;241;43m.[39;49m[43mclient[49m[38;5;241;43m.[39;49m[43mHTTPSConnection[49m[43m,[49m[43m [49m[43mreq[49m[43m,[49m
[1;32m   1392[0m [43m        [49m[43mcontext[49m[38;5;241;43m=[39;49m[38;5;28;43mself[39;49m[38;5;241;43m.[39;49m[43m_context[49m[43m,[49m[43m [49m[43mcheck_hostname[49m[38;5;241;43m=[39;49m[38;5;28;43mself[39;49m[38;5;241;43m.[39;49m[43m_check_hostname[49m[43m)[49m

File [0;32m~/miniconda3/envs/sta2003_nb/lib/python3.10/urllib/request.py:1351[0m, in [0;36mAbstractHTTPHandler.do_open[0;34m(self, http_class, req, **http_conn_args)[0m
[1;32m   1348[0m         h[38;5;241m.[39mrequest(req[38;5;241m.[39mget_method(), req[38;5;241m.[39mselector, req[38;5;241m.[39mdata, headers,
[1;32m   1349[0m                   encode_chunked[38;5;241m=[39mreq[38;5;241m.[39mhas_header([38;5;124m'[39m[38;5;124mTransfer-encoding[39m[38;5;124m'[39m))
[1;32m   1350[0m     [38;5;28;01mexcept[39;00m [38;5;167;01mOSError[39;00m [38;5;28;01mas[39;00m err: [38;5;66;03m# timeout error[39;00m
[0;32m-> 1351[0m         [38;5;28;01mraise[39;00m URLError(err)
[1;32m   1352[0m     r [38;5;241m=[39m h[38;5;241m.[39mgetresponse()
[1;32m   1353[0m [38;5;28;01mexcept[39;00m:

[0;31mURLError[0m: <urlopen error [Errno 8] nodename nor servname provided, or not known>

